### ISSUE TITLE
Unable to run SDL referenced code 

### ISSUE THREAD

**Issue by thirdknife** [mosra/magnum#241]
Cannot find ```#include <SDL.h>``` in ```SDLApplication.h```. SDL is installed correct.

Running on Ubuntu Mate 16.04 

Help is appreciated.

**Comment by Squareys** [comment]
Hi @thirdknife !

Is this while building magnum or while building your project? Which packages did you install for sdl?

Cheers, Jonathan.

**Comment by thirdknife** [comment]
@Squareys While building my project. 

I installed SDL from source. I also installed SDL image, mixed and ttf.

**Comment by thirdknife** [comment]
Demos which relies on SDL don't work either.

**Comment by mosra** [comment]
Hi!

some background on this: because not all platforms have SDL includes available in `<SDL2/SDL.h>` (for example macOS has the include in `SDL2.framework/Headers/SDL.h`), the `Sdl2Application.h` has to directly include `SDL.h` instead of `SDL2/SDL.h`. That might be the problem you are experiencing.

Are you using CMake? In that case you need to set `SDL_INCLUDE_DIR` to the `SDL/` subdirectory of where your SDL includes are. If not, you need to add the `SDL/` to your include path.

> I installed SDL from source.

Just a tip to save you some headaches: the `libsdl2-dev` package that comes with Ubuntu 16.04 "just works" and CMake detects it properly.

Hope this helps :)

**Comment by thirdknife** [comment]
Hi!

So I have dig deeper into the files and it seems like it does find right paths for SDL. But for your second part where you say set the `SDL_INCLUDE_DIR` I couldn't find a way to set it, there is none present in the CMake files. Can you please explain how to achieve that? 

cmake command looks like this ```cmake ..     -DCMAKE_INSTALL_PREFIX=/usr     -DWITH_SDL2APPLICATION=ON     -DSDL2_INCLUDE_DIR='/usr/local/include/SDL2'``` and below is my output is as:

```-- The CXX compiler identification is GNU 5.4.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Corrade: /usr/local/include  found components:  Containers rc Utility 
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libGL.so  
-- LIB_SUFFIX variable is not defined. It will be autodetected now.
-- You can set it manually with -DLIB_SUFFIX=<value> (64 for example)
-- LIB_SUFFIX autodetected as '', libraries will be installed into /usr/lib
[STATUS]sdl2 path......
[STATUS]SDL2_LIBRARY
[STATUS]/usr/local/include/SDL2
-- Found SDL2: /usr/lib/x86_64-linux-gnu/libSDL2-2.0.so  
-- Found Corrade: /usr/local/include  found components:  Containers rc Utility TestSuite 
-- Found Corrade: /usr/local/include  found components:  Containers rc Utility PluginManager 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/user/magnum/build```

**Comment by mosra** [comment]
Argh, sorry, it's `SDL2_INCLUDE_DIR`, not `SDL_INCLUDE_DIR`. Wrote that off the top of my head and was wrong. So the way you set it is correct.

So, with the above command it works? What `SDL2_INCLUDE_DIR` does it detect if you run cmake in a clean build directory without the parameter?

Something looks strange to me here: it finds the library in `/usr/lib`, but the includes are in `/usr/local/include/SDL`. That doesn't look like a standard installation to me.

**Comment by thirdknife** [comment]
I tried to do in a clean docker ubuntu container. cmake log for 

`cmake .. \
    -DCMAKE_INSTALL_PREFIX=/usr \
    -DWITH_SDL2APPLICATION=ON`

looks like this:

```
-- The CXX compiler identification is GNU 5.4.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Corrade: /usr/include  found components:  Containers rc Utility 
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libGL.so  
-- LIB_SUFFIX variable is not defined. It will be autodetected now.
-- You can set it manually with -DLIB_SUFFIX=<value> (64 for example)
-- LIB_SUFFIX autodetected as '', libraries will be installed into /usr/lib
-- Found SDL2: /usr/lib/x86_64-linux-gnu/libSDL2-2.0.so  
-- Found Corrade: /usr/include  found components:  Containers rc Utility TestSuite 
-- Found Corrade: /usr/include  found components:  Containers rc Utility PluginManager 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/magnum/build
```

When I try to build cubemap example I get the following CMake error:

```
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Corrade: /usr/include  found components:  Containers rc Utility PluginManager 
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libGL.so  
-- Found Magnum: /usr/include  found components:  Trade MeshTools Primitives SceneGraph Sdl2Application 
-- LIB_SUFFIX variable is not defined. It will be autodetected now.
-- You can set it manually with -DLIB_SUFFIX=<value> (64 for example)
-- LIB_SUFFIX autodetected as '', libraries will be installed into /usr/local/lib
-- Configuring done
CMake Error at CMakeLists.txt:63 (add_executable):
  Target "magnum-cubemap" links to target "SDL2::SDL2" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


-- Generating done
-- Build files have been written to: /home/magnum-examples/src/cubemap/build
```

**Comment by Squareys** [comment]
@thirdknife I will try to reproduce this in a VM later. If you could give me some addition info that would be great:

- Is there a specific reason why you cannot use the sdl package and use the source packages instead?
- I suppose you downloaded sdl source from the libsdl Website? Why command did you use to install it?
- Which cmake version are you using? Is it the package from the official repo or how did you install that?

Thanks in advance!
Cheers, Jonathan.

**Comment by mosra** [comment]
Was finally able to reproduce this on my Ubuntu VM and getting the same error as you, the `cd cubemap` part was the missing piece -- I didn't know you were building just a single example separately. That seems to be broken with CMake 3.5, on my main machine with 3.11 it seems to work. Will investigate further and fix this.

As a temporary solution, if you build the whole example repository like this (i.e., not just a single example), then it should work:

```sh
cd magnum-examples
mkdir build && cd build
cmake .. -DWITH_CUBEMAP_EXAMPLE=ON
make -j
```

**Comment by mosra** [comment]
Should be fixed in mosra/magnum-examples@a5144b2b71b58478eb48f3e7cde0ac9a1daf74bb. Turns out it was not due to CMake 3.5, but due to my system having a SDL installation that contains CMake config modules, so it "magically" found it even though `FindSDL2.cmake` was not available.

Please confirm this works for you now.

**Comment by mosra** [comment]
Great, happy to hear it's fixed :)
